### ISSUE TITLE
Seems like `sampleBurstEnd` has wrong implementation

### ISSUE THREAD

**Issue by TheKK** [composewell/streamly#1610]
https://github.com/composewell/streamly/blob/254ff5cc9956672edacf46851cc5d592f36551c7/src/Streamly/Internal/Data/Stream/IsStream/Top.hs#L153

I found that this `sampleBurstEnd` didn't work as I expected so I decide to investigate it a little. Here's its implementation:

```haskell
sampleBurstEnd :: (IsStream t, MonadAsync m, Functor (t m)) =>
    Double -> t m a -> t m a
sampleBurstEnd gap =
    let f (t1, _) (t2, _) =
            t2 - t1 >= toRelTime64 (NanoSecond64 (round (gap * 10^(9::Int))))
    in Stream.map snd
        . Stream.catMaybes
        . Stream.groupsByRolling f Fold.last
        . Stream.timeIndexed
```

The key here is `groupsByRolling`. According to documentation and my experimentation, when `f` returns `True` we'll push new value to the existed group and when `f` returns `False` we'll **seal** the old group and push new value to new group.

In that case, `f` should returns `True` when time stamp between two values are less than `gap`, and returns `False` when time stamp of two value are far enough (far than `gap`). That means `Stream.groupsByRolling (not . f) Fold.last` might be the correct implementation.

Also, I'm not sure if we should explain `sampleBurstEnd` as `debounce` here. Take the following graph as example:

```
1 (0s) --> 2 (1s) ---> 3 (4s) -->
```

With `sampleBurstEnd 2`, the value **2** will be fired at 4s, but in library like rxjs value **2** should be fired near 3s (1s + gap 2s). This also means that without value **3** at 4s, value **2** will never be fired, which is a little strange behaviour to me.

**Comment by harendra-kumar** [comment]
Yes, the implementation is not correct. Thanks for creating the issue. 

How about something like this?

```haskell
sampleBurstEnd :: (IsStream t, MonadAsync m) => Double -> t m a -> t m a
sampleBurstEnd gap xs =
    let ticks = gap / 0.01

        step (mx, count) Nothing = (mx, count + 1)
        step _ (Just x) = (Just x, 0)

        extract (Just x, count) | count >= ticks = Just x
        extract _ = Nothing

     in Stream.mapMaybe extract
        $ Stream.scanl' step (Nothing, 0)
        $ Stream.interjectSuffix 0.01 (return Nothing) (Stream.map Just xs)
```

It should fix both the issues you pointed out.

**Comment by harendra-kumar** [comment]
A better (more precise) fix is proposed in #1613 . Please take a look and let me know if it works as expected for you.
